tests/unit/src/test/scala/tests/debug/BreakpointDapSuite.scala

```diff
@@ -0,0 +1,468 @@
+                |  }
+                |}
+                |""".stripMargin
+  )
+
+  assertFileBreakpoints("nested class")(
+    source = """|a/src/main/scala/a/Main.scala
+                |package a
+                |
+                |trait Foo {
+                |  class Bar {
+                |    def call() = {
+                |>>    println()
+                |    }
+                |  }
+                |}
+                |
+                |object Main extends Foo {
+                |    def main(args: Array[String]): Unit = {
+                |      val bar = new Bar
+                |      bar.call()
+                |    }
+                |  }
+                |""".stripMargin
+  )
+
+  assertFileBreakpoints("nested trait")(
+    source = """|a/src/main/scala/a/Main.scala
+                |package a
+                |
+                |trait Foo {
+                |  trait Bar {
+                |    def call() = {
+                |>>    println()
+                |    }
+                |  }
+                |}
+                |
+                |object Main extends Foo {
+                |    def main(args: Array[String]): Unit = {
+                |      val bar = new Bar {}
+                |      bar.call()
+                |    }
+                |  }
+                |""".stripMargin
+  )
+
+  assertProjectBreakpoints("package-object")(
+    sources = List(
+      """|a/src/main/scala/a/package.scala
+         |package object a {
+         |  def call() = {
+         |>>  println()
+         |  }
+         |}
+         |""".stripMargin,
+      """|a/src/main/scala/a/Main.scala
+         |package a
+         |object Main {
+         |    def main(args: Array[String]): Unit = {
+         |      call()
+         |    }
+         |}
+         |""".stripMargin
+    )
+  )
+
+  assertProjectBreakpoints("not-matching-filename")(
+    sources = List(
+      """|a/src/main/scala/a/not-matching.scala
+         |package a
+         |
+         |object Foo {
+         |  def call() = {
+         |>>  println()
+         |  }
+         |}
+         |""".stripMargin,
+      """|a/src/main/scala/a/Main.scala
+         |package a
+         |object Main {
+         |    def main(args: Array[String]): Unit = {
+         |      a.Foo.call()
+         |    }
+         |}
+         |""".stripMargin
+    )
+  )
+
+  assertProjectBreakpoints("not-matching-package")(
+    sources = List(
+      """|a/src/main/scala/Foo.scala
+         |package not.matching
+         |
+         |object Foo {
+         |  def call() = {
+         |>>  println()
+         |  }
+         |}
+         |""".stripMargin,
+      """|a/src/main/scala/a/Main.scala
+         |package a
+         |object Main {
+         |    def main(args: Array[String]): Unit = {
+         |      not.matching.Foo.call()
+         |    }
+         |}
+         |""".stripMargin
+    )
+  )
+
+  // TODO OnDemandSymbolIndex doesn't distinguish between build targets
```

Comment:
This is a known limitation that also applies to "goto definition". We can open a followup ticket and remove this TODO

Reply:
Opened https://github.com/scalameta/metals/issues/1196